src/baldr/graphtile.cc

```diff
@@ -126,24 +129,51 @@ GraphTile::GraphTile(const GraphId& graphid, char* ptr, size_t size) : header_(n
   Initialize(graphid, ptr, size);
 }
 
-GraphTile::GraphTile(const std::string& tile_url, const GraphId& graphid, curler_t& curler) {
+GraphTile GraphTile::CacheTileURL(const std::string& tile_url,
+                                  const GraphId& graphid,
+                                  curler_t& curler,
+                                  bool gzipped,
+                                  const std::string& cache_location) {
   // Don't bother with invalid ids
   if (!graphid.Is_Valid() || graphid.level() > TileHierarchy::get_max_level()) {
-    return;
+    return {};
   }
 
-  // Get the response returned from curl
-  std::string uri =
-      tile_url + filesystem::path::preferred_separator + FileSuffix(graphid.Tile_Base());
+  // Get the response returned from curl, notice that when we expect zipped
+  // we tell curl not to unzip it, so that we can store it on the disk unzipped
+  auto suffix = FileSuffix(graphid.Tile_Base(), gzipped);
+  auto uri = tile_url + '/' + suffix;
   long http_code;
-  auto tile_data = curler(uri, http_code);
+  auto tile_data = curler(uri, http_code, false);
 
   // If its good try to use it
   if (http_code == 200) {
-    graphtile_ = std::make_shared<std::vector<char>>(std::move(tile_data));
-    Initialize(graphid, graphtile_->data(), graphtile_->size());
-    // TODO: optionally write the tile to disk?
+    // try to cache it on disk so we dont have to keep doing this
+    if (!cache_location.empty()) {
```

Comment:
the caching part is optional though, so if you dont specify a `tile_dir` in the valhalla config it will just use the in memory cache as it did before with urls